MODFLOW-USGS/modflow6
# refactor: simplify autotest framework

Changes include

* fold `TestSimulation` into `TestFramework`
* rename/refactor properties
  * `simpath` -> `workspace`
  * `build_function` -> `build` (and inject in `__init__`)
  * `exfunc` -> `check`
  * `exe_dict` -> `targets`
  * `require_failure` -> `xfail`
  * `action` -> `compare`
  * fold switches `run_comparison`, `make_comparison`, `mf6_regression` into `compare` (optional str), accepts:
    * "auto" picks comparison type based on workspace contents, equivalent to previous "compare" (default)
    * "run_only"  equivalent to previous `run_comparison=True and make_comparison=False`
    * "mf6_regression" equivalent to previous `mf6_regression=True`
    * `None` disables comparison
* standardize test function names, step toward automatic discovery?
  * `build_models` creates models
  * `check_output` checks results
* framework doesn't need to know about test case indices, remove `idxsim` and just use lambda if `build_models()` needs a test case index
* support `pathlib.Path` or `str` paths
* run black/isort, format test description docstrings
* remove unused functions from `common_regression.py`
* remove [`pytest-cases`](https://smarie.github.io/python-pytest-cases/) from tests and as a test dependency
* fix hardcoded gridgen left in `conftest.py` from #1460 

Merge considerations

- requires modflow-devtools >= 1.1 (update `environment.yml` to specify this)
- should probably wait for #1467 and any other open PRs touching autotests, so any conflicts can be resolved here

Future directions

- support multiple comparison/regression models? (e.g. compare current mf6 with previous mf6 and mf2005, or PRT with mp6 and mp7)
- explicit bookkeeping for sim/model workspaces? (replace implicit expectation that test and framework agree)
- autodiscovery of test cases following e.g. [this pattern](https://docs.pytest.org/en/7.1.x/example/nonpython.html#yaml-plugin)?
